zkml/src/tensor.rs

```diff
@@ -1054,52 +1054,83 @@ where
             }
         }
     }
-    pub fn pad_to_shape(&mut self, target_shape: Shape) {
+
+    /// Changes the shape of the cursort to `target_shape.`.
+    ///
+    /// This method will modify the current tensor in place, extending it
+    /// to comply with the new shape.
+    ///
+    /// # Panics
+    ///
+    /// If the `target_shape` differs in rank or has a dimension smaller than
+    /// the current tensor.
+    pub fn pad_to_shape_in_place(&mut self, target_shape: Shape) {
         assert!(
             target_shape.rank() == self.shape.rank(),
-            "Target shape must have the rank as the current tensor."
+            "Target shape must have the same rank as the current tensor."
         );
+
+        let distance = self
+            .shape
+            .iter()
+            .zip(target_shape.iter())
+            .map(|(original, new)| new.checked_sub(*original))
+            .collect::<Option<Vec<usize>>>();
+
         assert!(
-            self.shape
-                .iter()
-                .zip(target_shape.iter())
-                .all(|(c, t)| c <= t),
+            distance.is_some(),
             "All dimensions of target shape must be greater-than-or-equal to the current tensor",
         );
+        let distance = distance.unwrap();
 
-        let current_shape = &self.shape;
+        // First expand the underlying storage vector to the new size
+        self.data.resize(target_shape.product(), T::default());
```

Comment:
To reap the benefits, follow up changes will be necessary, changing the inputs parser to use `Vec::with_capacity` and pre-allocate the require space.

As it is, the usefulness of this change depends on a lot of factors, like the capacity of `self.data` the new sizes of `target_shape`.